### PR TITLE
When container is started in host networkMode, we mount a noop weavewait

### LINKED ISSUE #1600: Containers get stuck on weavewait when NetworkMode=host is provided at start
Specific case of https://github.com/weaveworks/weave/issues/1258 which is closed but may be worth reviewing (it was closed with *It's too intrusive a change for 1.0.2.*)

Spawning a container which provides `NetworkMode=host` at start results in weavewait waiting indefinitely.

To reproduce:

```
$ DOCKER_HOST=unix:///var/run/weave/weave.sock docker create --name foo busybox echo foo
$ curl -v --unix-socket /var/run/weave/weave.sock -XPOST http:/v1.15/containers/foo/start -H "Content-Type: application/json" --data-binary '{"HostConfig": { "NetworkMode": "host"} }'
$ ps auxf | grep 'echo foo'
root     16605  0.0  0.0   4880  1808 ?        Ssl  14:41   0:00  \_ /w/w echo foo
```

The weave proxy seems to recognize the host network mode ...
```
INFO: 2015/10/28 14:39:44.275603 POST /v1.15/containers/foo/start
INFO: 2015/10/28 14:39:44.343674 Leaving container c8af2e50a472eaf274b2e2fbd0b468debb814f4fde90634e161c78ad16ca71df alone because the container was created with the '--net=host'
```

... but it doesn't really *leave the container alone*.

In an offline discussion, @paulbellamy mentioned this happens because by the time the container starts it was already created with the weavewait entrypoint.

I experienced this when trying to run Kubernetes with Weave (which may not be a sane thing to do but that's besides the point).

Related: https://github.com/kubernetes/kubernetes/issues/12263

### PR DESCRIPTION
**Pull request by paulbellamy**
Fixes #1600

### CODE REVIEW

File: proxy/proxy.go [hunk 1]
```diff
@@ -424,6 +433,14 @@ func (proxy *Proxy) weaveCIDRs(networkMode string, env []string) ([]string, erro
 }
 
 func (proxy *Proxy) addWeaveWaitVolume(hostConfig jsonObject) error {
+	return proxy.addVolume(hostConfig, proxy.weaveWaitVolume, "/w", "ro")
+}
+
+func (proxy *Proxy) addWeaveWaitNoopVolume(hostConfig jsonObject) error {
+	return proxy.addVolume(hostConfig, proxy.weaveWaitNoopVolume, "/w", "ro")
+}
```
Comment (rade) [inline proxy/proxy.go]: ditch those functions.

File: test/666_abuse_of_start_test.sh [hunk 2]
```diff
@@ -36,12 +44,25 @@ assert "docker_on $HOST1 inspect -f '{{.HostConfig.MemorySwap}}' c4" "-1"
 # Start c6 with both named and unnamed HostConfig
 proxy docker_on $HOST1 create --name=c6 $SMALL_IMAGE $CHECK_ETHWE_UP
 proxy docker_api_on $HOST1 POST /containers/c6/start '{"NetworkMode": "container:c2", "HostConfig": {"NetworkMode": "container:c1"}}'
-check_hostconfig c6 container:c1
+check_dns c3
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: c6?

File: test/666_abuse_of_start_test.sh [hunk 3]
```diff
@@ -8,25 +8,33 @@ weave_on $HOST1 launch
...
+    docker_on $HOST1 attach c2 >/dev/null 2>&1 || true # Wait for container to exit
+    assert "docker_on $HOST1 inspect -f '{{.HostConfig.NetworkMode}} {{.State.Running}} {{.State.ExitCode}}' $1" "$2 false 0"
 }
 
 # Start c2 with a sneaky HostConfig
 proxy docker_on $HOST1 create --name=c2 $SMALL_IMAGE $CHECK_ETHWE_UP
 proxy docker_api_on $HOST1 POST /containers/c2/start '{"NetworkMode": "container:c1"}'
-check_hostconfig c2 container:c1
+check_dns c2
+check_networkmode c2 container:c1
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: It's annoying that this ends up doing two 'attach' and two 'inspect'. I'd just have a single function, with an optional param to indicate whether to check dns.

File: test/666_abuse_of_start_test.sh [hunk 4]
```diff
@@ -8,25 +8,33 @@ weave_on $HOST1 launch
 docker_bridge_ip=$(weave_on $HOST1 docker-bridge-ip)
 proxy_start_container $HOST1 --name=c1
 
-check_hostconfig() {
+check_dns() {
     docker_on $HOST1 attach c2 >/dev/null 2>&1 || true # Wait for container to exit
```
Comment (rade) [inline test/666_abuse_of_start_test.sh]: eek. that looks like a bug. should be `$1` instead of `c2`, surely.

**Comment by rade** [comment]
you accidentally checked in `prog/weavewait/weavewait_noop`.

File: proxy/proxy.go [hunk 5]
```diff
@@ -423,7 +432,7 @@ func (proxy *Proxy) weaveCIDRs(networkMode string, env []string) ([]string, erro
 	return nil, nil
 }
 
-func (proxy *Proxy) addWeaveWaitVolume(hostConfig jsonObject) error {
+func (proxy *Proxy) addVolume(hostConfig jsonObject, source, target, mode string) error {
```
Comment (rade) [inline proxy/proxy.go]: This should just be a function, not a method on `Proxy`.

File: proxy/proxy.go [hunk 6]
```diff
@@ -495,3 +508,14 @@ func (proxy *Proxy) getDNSDomain() (domain string) {
...
+func (proxy *Proxy) networkModeIsAttachable(networkMode string) (bool, error) {
+	for strings.HasPrefix(networkMode, "container:") {
+		container, err := proxy.client.InspectContainer(strings.TrimPrefix(networkMode, "container:"))
+		if err != nil || container.HostConfig == nil {
+			return false, err
+		}
+		networkMode = container.HostConfig.NetworkMode
+	}
+	return networkMode != "host", nil
+}
```
Comment (rade) [inline proxy/proxy.go]: There is a problem with this....

The parent container could be restarted with a different network mode. This is made worse by the fact that the reference can be a name, so you could actually change the parent completely.

My experiments indicate that this does not affect a running container. But does when the container is restarted. Which, if it happens w/o going through the proxy, e.g. due a restart policy, means we can end up with the wrong weavewait. Not much we can do about that, I suppose, but probably merits a comment.

Another edge case is that the parent container may not exist. Looks like we error at that point; shouldn't we leave that to docker? Though it is of course possible that the parent may appear by the time the request reaches docker. hmm.

**Comment by rade** [comment]
When we deal with network modes at container creation time, we do not insert a weavewait entrypoint when the mode is `container:...`. Even though that container could be weave networked.

Here, at start, we mount weavewait (instead of weavewait-noop) when the mode is `container:...` unless the parent is on the host network.

That is inconsistent.

Ideally in both cases we'd want to weavewait if the parent container is weave networked. But we have no reliable way of figuring that out, I think.

So perhaps our logic should simply be, "if the network mode is `host` or `container:...`" then don't weavewait.

**Comment by rade** [comment]
Another problem: the volume mount swivelling won't alter the entrypoint, which is all the proxy looks at when deciding whether to attach. Hmm. Except it also looks at the NetworkMode. arghhh, this is all hideously complicated.

File: proxy/proxy.go (hunk 6 above)
Comment (paulbellamy) [inline proxy/proxy.go]: A restart policy won't change the networkMode, though. Only an api stop/start with the hostconfig provided.

We could do these checks every time a container is started, and send a hostConfig if needed (regardless of whether the user sent one).

**Comment by paulbellamy** [comment]
per f2f discussion:

I agree we should simplify the logic to just not weavewait if `NetworkMode` is `host` or `container:...`

**Comment by paulbellamy** [comment]
Eugh, but that also means we won't set up any dns on these container either...

**Comment by rade** [comment]
anything left to do here? The failures should get cured with a rebase.

**Comment by paulbellamy** [comment]
Nope, should be ready to go.

**Pull request merged**